File: java/algorithm/string/leetcode/ReverseOnlyLetters.java
package algorithm.string.leetcode;

/**
 * 仅仅翻转字母
 */
public class ReverseOnlyLetters {

    /**
     * 双指针
     */
    public String reverseOnlyLetters(String S) {
        char[] chars = S.toCharArray();
        int i = 0, j = chars.length - 1;
        while (i < j) {
            if (!Character.isLetter(chars[i])) {
                i++;
            } else if (!Character.isLetter(chars[j])) {
                j--;
            } else {
                char temp = chars[i]; chars[i] = chars[j]; chars[j] = temp;
                i++; j--;
            }
        }
        return String.valueOf(chars);
    }
}

File: go/string/reverse_only_letters.go
package string

import (
	"unicode"
)

// 仅仅翻转字母
func reverseOnlyLetters(S string) string {
	runes := []rune(S)
	i, j := 0, len(runes)-1
	for i < j {
		if !unicode.IsLetter(runes[i]) {
			i++
		} else if !unicode.IsLetter(runes[j]) {
			j--
		} else {
			runes[i], runes[j] = runes[j], runes[i]
			i++
			j--
		}
	}
	return string(runes)
}
